File: api/src/main/java/net/asch/bulkit/api/registry/DeferredResources.java
package net.asch.bulkit.api.registry;

import net.asch.bulkit.api.BulkIt;
import net.minecraft.core.Registry;
import net.minecraft.resources.ResourceKey;
import net.neoforged.neoforge.registries.DeferredHolder;
import net.neoforged.neoforge.registries.DeferredRegister;
import net.neoforged.neoforge.registries.NewRegistryEvent;
import net.neoforged.neoforge.registries.RegistryBuilder;

import java.util.Collection;

public class DeferredResources extends DeferredRegister<ResourceType<?>> {
    private static final ResourceKey<Registry<ResourceType<?>>> REGISTRY_KEY = ResourceKey.createRegistryKey(BulkIt.location("resource_types"));
    public static final Registry<ResourceType<?>> REGISTRY = new RegistryBuilder<ResourceType<?>>(REGISTRY_KEY).sync(true).create();

    public DeferredResources(String namespace) {
        super(REGISTRY_KEY, namespace);
    }

    public <T> DeferredHolder<ResourceType<?>, ResourceType<T>> registerResourceType(ResourceType.Builder<T> builder) {
        return register(builder.key, builder);
    }

    public static void register(NewRegistryEvent event) {
        event.register(REGISTRY);
    }

    public static Collection<ResourceType<?>> registeredResources() {
        return REGISTRY.stream().toList();
    }
}

File: api/src/main/kotlin/net/asch/bulkit/api/resource/DeferredResources.kt
package net.asch.bulkit.api.resource

import net.asch.bulkit.api.BulkItApi
import net.minecraft.core.Registry
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.neoforged.neoforge.registries.DeferredHolder
import net.neoforged.neoforge.registries.DeferredRegister
import net.neoforged.neoforge.registries.NewRegistryEvent
import net.neoforged.neoforge.registries.RegistryBuilder

class DeferredResources private constructor(modId: String) : DeferredRegister<ResourceType<*>>(REGISTRY_KEY, modId) {
    fun <R> registerResourceType(builder: ResourceType.Builder<R>): DeferredHolder<ResourceType<*>, ResourceType<R>> =
        register(builder.key, builder)

    companion object {
        private val REGISTRY_KEY: ResourceKey<Registry<ResourceType<*>>> =
            ResourceKey.createRegistryKey(ResourceLocation.fromNamespaceAndPath(BulkItApi.ID, "resource_types"))
        private val REGISTRY: Registry<ResourceType<*>> = RegistryBuilder(REGISTRY_KEY).sync(true).create()

        fun onNewRegistry(event: NewRegistryEvent) = event.register(REGISTRY)
        fun create(modId: String): DeferredResources = DeferredResources(modId)
        fun registeredResourceTypes(): Collection<ResourceType<*>> = REGISTRY.stream().toList()
    }
}